c10/cuda/CUDAAssert.h

```diff
@@ -0,0 +1,32 @@
+#pragma once
+
+#include <cstdint>
+
+namespace c10 { namespace cuda {
+
+enum class CUDAAssertKind : int32_t {
+  ASSERTION_FAILED,
+  INDEX_OUT_OF_BOUNDS,
+  ZERO_DIVISION,
+};
+
+constexpr int MAX_ASSERT_MSG_LENGTH = 1024;
+constexpr int MAX_ERROR_DETAILS_SIZE = 1024;
+
+struct CUDAAssert {
+  volatile int32_t error;
+  CUDAAssertKind type;
+  uint32_t line;
+  uint32_t _reserved;  // round to 8 byte boundary
+  char message[MAX_ASSERT_MSG_LENGTH];
+  char file[MAX_ASSERT_MSG_LENGTH];
+  int8_t details[MAX_ERROR_DETAILS_SIZE];
```

Comment:
You'll likely define all the detail structs in this file as well, so you can turn this into a union of those structs to avoid accidentally overflowing and excessive reinterpret_cast'ing.

Reply:
I thought about using an union. My first concern was that c10 should not be changed whenever new/special assert functionality is added, e.g. originally I wanted to create something like a registry for formatters/handlers. But if there are only a handful of different cases and we basically 'seal' the set of assert/error functions for users, it might be reasonable to specify them in a c10 header. I guess that would also solve alignment issues, e.g. currently if the `_reserved` member would be removed the field-writes would trigger 'misaligned address' errors in the `graceful_index_error_`...